File: ruby/nth-prime/nth_prime.rb
module Prime
  module IntegerHelpers
    refine Integer do
      def prime?
        (2..(self / 2)).each do |n|
          return false if (self % n).zero?
        end

        true
      end
    end
  end

  module EnumeratorHelpers
    refine Enumerator do
      def nth(n)
        (n-1).times { self.next }
        self.peek
      end
    end
  end

  using IntegerHelpers
  using EnumeratorHelpers

  def self.nth(n)
    raise ArgumentError.new if n <= 0

    primes.nth(n)
  end

  def self.primes
    Enumerator.new do |p|
      (2..).each { |current| p.yield(current) if current.prime? }
    end
  end
end

File: javascript/nth-prime/nth-prime.js
export class Prime {
  constructor() {
    this.primes = [2];
  }

  nth(n) {
    if (n <= 0) {
      throw new Error("Prime is not possible");
    }

    while (this.primes.length < n) {
      this.calculateNextPrime();
    }

    return this.primes[n - 1];
  }

  calculateNextPrime() {
    let value = this.lastCalculatePrime;

    while (true) {
      let isPrime = true;
      value++;

      for (const prime of this.primes) {
        if (value % prime === 0) {
          isPrime = false;
          break;
        }
      }

      if (isPrime) {
        this.primes.push(value);
        break;
      }
    }
  }

  get lastCalculatePrime() {
    return this.primes[this.primes.length - 1];
  }
}
